client/checkout/blocks/upe-deferred-intent-creation/payment-processor.js

```diff
@@ -0,0 +1,204 @@
+	const stripe = useStripe();
+	const elements = useElements();
+	const [ isUPEComplete, setIsUPEComplete ] = useState( false );
+	const [ selectedUPEPaymentType, setSelectedUPEPaymentType ] = useState(
+		''
+	);
+
+	const paymentMethodsConfig = getUPEConfig( 'paymentMethodsConfig' );
+	const isTestMode = getUPEConfig( 'testMode' );
+	const testingInstructionsIfAppropriate = isTestMode
+		? testingInstructions
+		: '';
+	const gatewayConfig = getPaymentMethods()[ upeMethods[ paymentMethodId ] ];
+	const billingData = useCustomerData().billingAddress;
+
+	useEffect(
+		() =>
+			onPaymentSetup( () => {
+				async function handlePaymentProcessing() {
+					if (
+						upeMethods[ paymentMethodId ] !== activePaymentMethod
+					) {
+						return;
+					}
+
+					if ( ! isUPEComplete ) {
+						return {
+							type: 'error',
+							message: __(
+								'Your payment information is incomplete.',
+								'woocommerce-payments'
+							),
+						};
+					}
+
+					if ( errorMessage ) {
+						return {
+							type: 'error',
+							message: errorMessage,
+						};
+					}
+
+					if (
+						gatewayConfig.supports.showSaveOption &&
+						shouldSavePayment &&
+						! paymentMethodsConfig[ selectedUPEPaymentType ]
+							.isReusable
+					) {
+						return {
+							type: 'error',
+							message:
+								'This payment method cannot be saved for future use.',
+						};
+					}
+
+					await validateElements( elements );
+
+					const paymentMethodObject = await api
+						.getStripe()
+						.createPaymentMethod( {
+							elements,
+							params: {
+								billing_details: getBillingDetails(
+									billingData
+								),
+							},
+						} );
+
+					return {
+						type: 'success',
+						meta: {
+							paymentMethodData: {
+								payment_method: paymentMethodId,
+								'wcpay-payment-method':
+									paymentMethodObject.paymentMethod.id,
+								'wcpay-fraud-prevention-token': getFraudPreventionToken(),
+								'wcpay-fingerprint': fingerprint,
+							},
+						},
+					};
+				}
+				return handlePaymentProcessing();
+			} ),
+		[
+			activePaymentMethod,
+			api,
+			elements,
+			fingerprint,
+			gatewayConfig,
+			paymentMethodId,
+			paymentMethodsConfig,
+			selectedUPEPaymentType,
+			shouldSavePayment,
+			upeMethods,
+			errorMessage,
+			onPaymentSetup,
+			billingData,
+			isUPEComplete,
+		]
+	);
+
+	usePaymentCompleteHandler(
+		api,
+		stripe,
+		elements,
+		onCheckoutAfterProcessingWithSuccess,
+		emitResponse,
+		shouldSavePayment
+	);
+
+	// Checks whether there are errors within a field, and saves them for later reporting.
+	const upeOnChange = ( event ) => {
+		// Update WC Blocks gateway config based on selected UPE payment method.
+		const paymentType =
+			'link' !== event.value.type ? event.value.type : 'card';
+		gatewayConfig.supports.showSaveOption =
+			paymentMethodsConfig[ paymentType ].showSaveOption;
+
+		setIsUPEComplete( event.complete );
+		setSelectedUPEPaymentType( paymentType );
+	};
```

Comment:
I've checked all the values, and here's the result: 
1. `gatewayConfig.supports.showSaveOption` doesn't need to be set here since [the gateway config already comes with the relevant data for this setting](https://github.com/Automattic/woocommerce-payments/blob/7c5b120eaa82d82c58c0e883a7e93673ef92920a/client/checkout/blocks/upe-deferred-intent-creation/payment-processor.js#L74) ✔️ 
2. `selectedUPEPaymentType` indeed is not a must, mainly because [the `isReusable` setting can be easily fetched here by using `paymentMethodId` instead of `selectedUPEPaymentType`](https://github.com/Automattic/woocommerce-payments/blob/7c5b120eaa82d82c58c0e883a7e93673ef92920a/client/checkout/blocks/upe-deferred-intent-creation/payment-processor.js#L107-L108) ✔️ 

Having said that, I agree only `isUPEComplete` can be set within this function. I've adjusted the code in cc99e6a35aa468b84df4fd6cb3582bc7939e9696, where I also changed the function names to reflect the actual use case more precisely now. The code is shorter and easier to read now - thanks!